File: ruby/three_number_sum.rb
def three_number_sum(arr,target)

    j = 0
    k = arr.length - 1
    triple_pair = []

    for i in 0..arr.length - 2

        while j < k

            sum = arr[i] + arr[j] + arr[k]

            if sum == target
                triple_pair.push(arr[i],arr[j],arr[k])
                j+=1
                k-=1
                
            elsif sum > target
                k -= 1
            else
                j += 1
            end

        end

    end

    return triple_pair
end

# Driver code
arr = [1,2,3,4,5,6,7,8,9,10]
target = 10
p three_number_sum(arr,target)

File: rust/src/three_number_sum.rs
pub fn three_number_sum(arr: &mut [i32], target_sum: i32) -> Vec<[i32; 3]> {
    arr.sort();
    
    let mut triple_sum_arr = Vec::new();

    for i in 0..arr.len() - 2 {

        let mut left = i + 1;
        let mut right = arr.len() - 1;

        while left < right {
            let sum = arr[i] + arr[left] + arr[right];
            if sum == target_sum {
                triple_sum_arr.push([arr[i], arr[left], arr[right]]);
                left += 1;
                right -= 1;
            } else if sum < target_sum {
                left += 1;
            } else {
                right -= 1;
            }
        }
    }

    return triple_sum_arr
}